src/app.rs

```diff
@@ -125,88 +114,130 @@ impl App {
+                );
+                panic!("Event leaf does not match expected leaf.");
+            }
+
+            // Insert
+            merkle_tree.set(leaf, hash);
             next_leaf = max(next_leaf, leaf + 1);
+
+            // Check root
+            if root != merkle_tree.root() {
+                error!(computed_root = ?merkle_tree.root(), event_root = ?root, "Root mismatch between event and computed tree.");
+                panic!("Root mismatch between event and computed tree.");
+            }
         }
 
-        let is_manager = ethereum.is_manager().await?;
-        let nonce_offset = ethereum.get_nonce().await? - next_leaf;
-        info!(
-            "current nonce: {}, nonce offset: {}",
-            ethereum.get_nonce().await?,
-            nonce_offset
-        );
+        // TODO: Final root check
 
         Ok(Self {
             ethereum,
+            contracts,
             storage_file: options.storage_file,
             merkle_tree: RwLock::new(merkle_tree),
             next_leaf: AtomicUsize::new(next_leaf),
-            tree_depth: options.tree_depth,
-            is_manager,
-            nonce_offset,
         })
     }
 
     /// # Errors
     ///
     /// Will return `Err` if the Eth handler cannot insert the identity to the
     /// contract, or if writing to the storage file fails.
-    #[instrument(skip_all)]
+    #[instrument(level = "debug", skip_all)]
     pub async fn insert_identity(
         &self,
         group_id: usize,
         commitment: &Hash,
     ) -> Result<IndexResponse, ServerError> {
-        // Check if manager
-        if !self.is_manager {
-            return Err(ServerError::NotManager);
+        if U256::from(group_id) != self.contracts.group_id() {
+            return Err(ServerError::InvalidGroupId);
         }
 
+        // Get a lock on the tree for the duration of this operation.
+        // OPT: Sequence operations and allow concurrent inserts / transactions.
+        let mut tree = self.merkle_tree.write().await;
+
         // Fetch next leaf index
         let identity_index = self.next_leaf.fetch_add(1, Ordering::AcqRel);
 
-        // Send Semaphore transaction, nonce calculated to ensure ordering
-        self.ethereum
-            .insert_identity(
-                group_id,
-                commitment,
-                self.tree_depth,
-                identity_index + self.nonce_offset,
-            )
-            .await?;
+        // Send Semaphore transaction
+        self.contracts.insert_identity(commitment).await?;
 
         // Update and write merkle tree
-        {
-            let mut merkle_tree = self.merkle_tree.write().await;
-            merkle_tree.set(identity_index, *commitment);
+        tree.set(identity_index, *commitment);
+
+        // Downgrade write lock to read lock
+        let tree = tree.downgrade();
+
+        // Check tree root
+        if let Err(error) = self.contracts.assert_valid_root(tree.root()).await {
```

Comment:
this will also fail if the root is expired: https://github.com/worldcoin/world-id-example-airdrop/blob/03de53d2cb016ddef28b26e8237e85b62ec385c7/src/Semaphore.sol#L143

then you can never insert an element again. We should prob introduce a separate error for that in the contracts. I would just drop the `assert_valid_root` check in this get handler to be backwards compatible with our existing deployment. Also, it feels not super necessary.

Reply:
The root should not have expired because it just got updated on line 204. This is the new root just after insertion.

That being said, the root _before_ insertion should also be valid because it should be the latest root. I think the contract might be wrong here.

Reply:
oh right, actually not an issue. I saw the error originally on the `/inclusionProof` endpoint. No issue here for insertion.